File: src/hmmer.cpp
#include <string>
#include <map>
#include "hmmer.h"
#include "regexmacros.h"
#include "util.h"

using namespace MachineBoss;

double HmmerModel::strToProb (const string& s) {
  return s == "*" ? 0 : exp(-stof(s));
}

vguard<double> HmmerModel::strsToProbs (const vguard<string>& s) {
  vguard<double> p;
  p.reserve (s.size());
  for (auto& str: s)
    p.push_back (strToProb (str));
  return p;
}

void HmmerModel::loadNullModel() {
  // SwissProt background amino acid frequencies (from data/SwissProtComposition.json)
  const map<string, double> bgFreq = {
    {"A", 0.0825}, {"C", 0.0138}, {"D", 0.0546}, {"E", 0.0673},
    {"F", 0.0386}, {"G", 0.0708}, {"H", 0.0227}, {"I", 0.0592},
    {"K", 0.0581}, {"L", 0.0965}, {"M", 0.0241}, {"N", 0.0405},
    {"P", 0.0473}, {"Q", 0.0393}, {"R", 0.0553}, {"S", 0.0663},
    {"T", 0.0535}, {"V", 0.0686}, {"W", 0.0109}, {"Y", 0.0292}
  };
  nullEmit.clear();
  nullEmit.reserve (alph.size());
  for (const auto& sym : alph) {
    auto it = bgFreq.find (sym);
    if (it != bgFreq.end())
      nullEmit.push_back (it->second);
    else
      nullEmit.push_back (1.0 / alph.size());  // uniform fallback for unknown symbols
  }
}

void HmmerModel::read (istream& in) {
  node.clear();
  const regex tag_re ("^(" RE_PLUS(RE_CHAR_CLASS("A-Z")) ")");
  const regex end_re ("^//");
  string line;
  smatch tag_match;
  while (getline(in,line))
    if (regex_search (line, tag_match, tag_re)) {
      const string tag = tag_match.str(1);
      if (tag == "HMM") {
	const vguard<string> hmm_alph = split (line);
	Assert (hmm_alph.size() > 1, "HMM parse error: empty alphabet");
	alph = vguard<string> (hmm_alph.begin() + 1, hmm_alph.end());
	for (int skip = 0; skip < 3; ++skip)
	  if (!getline(in,line))
	    break;
	const vguard<string> ins0 = split (line);
	Assert (ins0.size() == alph.size(), "HMM parse error: wrong number of fields in node 0 insert line");
	ins0Emit = strsToProbs (ins0);
 	if (!getline(in,line))
	  break;
	const vguard<string> beginTrans = split (line);
	b_to_m1 = strToProb (beginTrans[0]);
	b_to_i0 = strToProb (beginTrans[1]);
	b_to_d1 = strToProb (beginTrans[2]);
	i0_to_m1 = strToProb (beginTrans[3]);
	i0_to_i0 = strToProb (beginTrans[4]);
	while (getline(in,line)) {
	  if (regex_match (line, end_re))
	    break;
	  const vguard<string> nodeMatchLine = split (line);
	  Assert (nodeMatchLine.size() == alph.size() + 6, "HMM parse error: wrong number of fields in node match line");
	  Assert (stoi(nodeMatchLine[0]) == node.size() + 1, "HMM parse error: incorrect node index");
	  Assert (getline(in,line), "HMM parse error: premature truncation of node after match line");
	  const vguard<string> nodeInsEmit = split (line);
	  Assert (nodeInsEmit.size() == alph.size(), "HMM parse error: wrong number of fields in node insert line");
	  Assert (getline(in,line), "HMM parse error: premature truncation of node after insert line");
	  const vguard<string> nodeTrans = split (line);
	  Assert (nodeTrans.size() == 7, "HMM parse error: wrong number of fields in node transitions line");
	  Node n;
	  const vguard<string> nodeMatchEmit (nodeMatchLine.begin() + 1, nodeMatchLine.begin() + alph.size() + 1);
	  n.matchEmit = strsToProbs (nodeMatchEmit);
	  n.insEmit = strsToProbs (nodeInsEmit);
	  n.m_to_m = strToProb (nodeTrans[0]);
	  n.m_to_i = strToProb (nodeTrans[1]);
	  n.m_to_d = strToProb (nodeTrans[2]);
	  n.i_to_m = strToProb (nodeTrans[3]);
	  n.i_to_i = strToProb (nodeTrans[4]);
	  n.d_to_m = strToProb (nodeTrans[5]);
	  n.d_to_d = strToProb (nodeTrans[6]);
	  node.push_back (n);
	}
	break;
      }
    }
  loadNullModel();
}

Machine HmmerModel::machine (bool local) const {
  Assert (node.size() > 0, "Attempt to create a transducer from an empty HMMER model");

  Machine m;
  m.state = vguard<MachineState> (nCoreStates());

  m.state[b_idx()].name = "B";
  if (local) {
    // local mode entry probabilities from p7_ProfileConfig() in HMMER3 source code
    const auto occ = calcMatchOccupancy();
    double Z = 0;
    for (int k = 1; k < node.size(); ++k)
      Z += occ[k] * (node.size() - k + 1);
    for (int k = 1; k < node.size(); ++k)
      m.state[b_idx()].trans.push_back (MachineTransition (string(), string(), m_idx(k), occ[k] / Z));
  } else {
    m.state[b_idx()].trans.push_back (MachineTransition (string(), string(), m_idx(1), b_to_m1));
    m.state[b_idx()].trans.push_back (MachineTransition (string(), string(), i_idx(0), b_to_i0));
    m.state[b_idx()].trans.push_back (MachineTransition (string(), string(), d_idx(1), b_to_d1));
  }

  m.state[ix_idx(0)].trans.push_back (MachineTransition (string(), string(), m_idx(1), i0_to_m1));
  m.state[ix_idx(0)].trans.push_back (MachineTransition (string(), string(), i_idx(0), i0_to_i0));

  for (size_t sym = 0; sym < alph.size(); ++sym)
    m.state[i_idx(0)].trans.push_back (MachineTransition (string(), alph[sym], ix_idx(0), ins0Emit[sym]));

  for (int n = 0; n <= node.size(); ++n) {
    const string ns = to_string(n);
    m.state[i_idx(n)].name = string("I") + ns;
    m.state[ix_idx(n)].name = string("Ix") + ns;
    if (n > 0) {
      m.state[m_idx(n)].name = string("M") + ns;
      m.state[mx_idx(n)].name = string("Mx") + ns;
      m.state[d_idx(n)].name = string("D") + ns;

      const bool end = (n == node.size());
      if (end) {
	if (!local)
	  m.state[mx_idx(n)].trans.push_back (MachineTransition (string(), string(), core_end_idx(), node[n-1].m_to_m));
      } else
	m.state[mx_idx(n)].trans.push_back (MachineTransition (string(), string(), m_idx(n+1), node[n-1].m_to_m));
      m.state[mx_idx(n)].trans.push_back (MachineTransition (string(), string(), i_idx(n), node[n-1].m_to_i));
      if (!end)
	m.state[mx_idx(n)].trans.push_back (MachineTransition (string(), string(), d_idx(n+1), node[n-1].m_to_d));

      m.state[ix_idx(n)].trans.push_back (MachineTransition (string(), string(), end ? core_end_idx() : m_idx(n+1), node[n-1].i_to_m));
      m.state[ix_idx(n)].trans.push_back (MachineTransition (string(), string(), i_idx(n), node[n-1].i_to_i));

      if (end) {
	if (!local)
	  m.state[d_idx(n)].trans.push_back (MachineTransition (string(), string(), core_end_idx(), node[n-1].d_to_m));
      } else {
	m.state[d_idx(n)].trans.push_back (MachineTransition (string(), string(), m_idx(n+1), node[n-1].d_to_m));
	m.state[d_idx(n)].trans.push_back (MachineTransition (string(), string(), d_idx(n+1), node[n-1].d_to_d));
      }

      for (size_t sym = 0; sym < alph.size(); ++sym) {
	m.state[m_idx(n)].trans.push_back (MachineTransition (string(), alph[sym], mx_idx(n), node[n-1].matchEmit[sym]));
	m.state[i_idx(n)].trans.push_back (MachineTransition (string(), alph[sym], ix_idx(n), node[n-1].insEmit[sym]));
      }

      if (local) {
	// HMMER3 allows unit-weight transitions from Match and Delete states to End state when in local mode, per p7_profile_GetT()
	m.state[m_idx(n)].trans.push_back (MachineTransition (string(), string(), core_end_idx(), WeightAlgebra::one()));
	m.state[d_idx(n)].trans.push_back (MachineTransition (string(), string(), core_end_idx(), WeightAlgebra::one()));
      }
    }
  }
  m.state[core_end_idx()].name = "E";

  return m;
}

Machine HmmerModel::plan7Machine (bool multihit, double L) const {
  Assert (node.size() > 0, "Attempt to create a Plan7 transducer from an empty HMMER model");
  Assert (nullEmit.size() == alph.size(), "Null model not loaded; call read() or loadNullModel() first");

  // Build core machine in local mode
  Machine core = machine (true);

  // Create Plan7 machine with extended state vector
  Machine m;
  m.state.resize (nPlan7States());

  // Copy core states (indices 0..nCoreStates()-1)
  for (StateIndex i = 0; i < nCoreStates(); ++i)
    m.state[i] = core.state[i];

  // Move B's original transitions to the Plan7 Begin state
  m.state[plan7_b_idx()] = m.state[b_idx()];
  m.state[plan7_b_idx()].name = "B";

  // Repurpose index 0 as S (Plan7 start)
  m.state[b_idx()].name = "S";
  m.state[b_idx()].trans.clear();
  m.state[b_idx()].trans.push_back (MachineTransition (string(), string(), nx_idx(), 1.0));

  // N-terminal flank: N emits background, Nx routes
  m.state[n_idx()].name = "N";
  for (size_t sym = 0; sym < alph.size(); ++sym)
    m.state[n_idx()].trans.push_back (MachineTransition (string(), alph[sym], nx_idx(), nullEmit[sym]));

  m.state[nx_idx()].name = "Nx";
  m.state[nx_idx()].trans.push_back (MachineTransition (string(), string(), n_idx(), L / (L + 1)));
  m.state[nx_idx()].trans.push_back (MachineTransition (string(), string(), plan7_b_idx(), 1.0 / (L + 1)));

  // E transitions to C-terminal flank (and optionally J for multi-hit)
  if (multihit) {
    m.state[core_end_idx()].trans.push_back (MachineTransition (string(), string(), cx_idx(), 0.5));
    m.state[core_end_idx()].trans.push_back (MachineTransition (string(), string(), jx_idx(), 0.5));
  } else {
    m.state[core_end_idx()].trans.push_back (MachineTransition (string(), string(), cx_idx(), 1.0));
  }

  // C-terminal flank: C emits background, Cx routes
  m.state[c_idx()].name = "C";
  for (size_t sym = 0; sym < alph.size(); ++sym)
    m.state[c_idx()].trans.push_back (MachineTransition (string(), alph[sym], cx_idx(), nullEmit[sym]));

  m.state[cx_idx()].name = "Cx";
  m.state[cx_idx()].trans.push_back (MachineTransition (string(), string(), c_idx(), L / (L + 1)));
  m.state[cx_idx()].trans.push_back (MachineTransition (string(), string(), t_idx(), 1.0 / (L + 1)));

  // J loop (multi-hit): J emits background, Jx routes back to B
  m.state[j_idx()].name = "J";
  m.state[jx_idx()].name = "Jx";
  if (multihit) {
    for (size_t sym = 0; sym < alph.size(); ++sym)
      m.state[j_idx()].trans.push_back (MachineTransition (string(), alph[sym], jx_idx(), nullEmit[sym]));

    m.state[jx_idx()].trans.push_back (MachineTransition (string(), string(), j_idx(), L / (L + 1)));
    m.state[jx_idx()].trans.push_back (MachineTransition (string(), string(), plan7_b_idx(), 1.0 / (L + 1)));
  }

  // Terminal state (machine end, no transitions)
  m.state[t_idx()].name = "T";

  return m;
}

vguard<double> HmmerModel::calcMatchOccupancy() const {
  // Taken from p7_hmm_CalculateOccupancy() in HMMER3 source code:
  //   Calculates a vector <mocc[1..M]> containing probability
  //   that each match state is used in a sampled path through
  //   the model.
  vguard<double> mocc (node.size());
  mocc[0] = 0.;			               /* no M_0 state */
  mocc[1] = node[0].m_to_i + node[0].m_to_m;   /* initialize w/ 1 - B->D_1 */
  for (int k = 2; k < node.size(); k++)
    mocc[k] = mocc[k-1] * (node[k].m_to_m + node[k].m_to_i) + (1.0 - mocc[k-1]) * node[k].d_to_m;
  return mocc;
}

File: python/machineboss/hmmer.py
"""Pure Python HMMER3 file parser.

Mirrors the C++ HmmerModel class, producing Machine objects.
"""

from __future__ import annotations

import math
import re
from dataclasses import dataclass, field
from typing import TextIO

from .machine import Machine, MachineState, MachineTransition

# SwissProt background amino acid frequencies
SWISSPROT_BG: dict[str, float] = {
    "A": 0.0825, "C": 0.0138, "D": 0.0546, "E": 0.0673,
    "F": 0.0386, "G": 0.0708, "H": 0.0227, "I": 0.0592,
    "K": 0.0581, "L": 0.0965, "M": 0.0241, "N": 0.0405,
    "P": 0.0473, "Q": 0.0393, "R": 0.0553, "S": 0.0663,
    "T": 0.0535, "V": 0.0686, "W": 0.0109, "Y": 0.0292,
}


@dataclass
class HmmerNode:
    match_emit: list[float] = field(default_factory=list)
    ins_emit: list[float] = field(default_factory=list)
    m_to_m: float = 0.0
    m_to_i: float = 0.0
    m_to_d: float = 0.0
    i_to_m: float = 0.0
    i_to_i: float = 0.0
    d_to_m: float = 0.0
    d_to_d: float = 0.0


@dataclass
class HmmerModel:
    """HMMER3 profile HMM model."""
    alph: list[str] = field(default_factory=list)
    nodes: list[HmmerNode] = field(default_factory=list)
    ins0_emit: list[float] = field(default_factory=list)
    null_emit: list[float] = field(default_factory=list)
    b_to_m1: float = 0.0
    b_to_i0: float = 0.0
    b_to_d1: float = 0.0
    i0_to_m1: float = 0.0
    i0_to_i0: float = 0.0

    @staticmethod
    def str_to_prob(s: str) -> float:
        return 0.0 if s == "*" else math.exp(-float(s))

    @classmethod
    def read(cls, f: TextIO) -> HmmerModel:
        """Parse an HMMER3 format file."""
        model = cls()
        tag_re = re.compile(r"^([A-Z]+)")
        end_re = re.compile(r"^//")

        for line in f:
            m = tag_re.match(line)
            if m and m.group(1) == "HMM":
                tokens = line.split()
                assert len(tokens) > 1, "HMM parse error: empty alphabet"
                model.alph = tokens[1:]

                # Skip transition header line, COMPO line, node 0 insert emission line
                for _ in range(3):
                    line = next(f)

                # Node 0 insert emissions
                ins0 = line.split()
                assert len(ins0) == len(model.alph)
                model.ins0_emit = [cls.str_to_prob(s) for s in ins0]

                # Begin transitions
                line = next(f)
                bt = line.split()
                model.b_to_m1 = cls.str_to_prob(bt[0])
                model.b_to_i0 = cls.str_to_prob(bt[1])
                model.b_to_d1 = cls.str_to_prob(bt[2])
                model.i0_to_m1 = cls.str_to_prob(bt[3])
                model.i0_to_i0 = cls.str_to_prob(bt[4])

                # Parse nodes
                for line in f:
                    if end_re.match(line):
                        break
                    match_fields = line.split()
                    assert len(match_fields) == len(model.alph) + 6

                    line = next(f)  # insert emission line
                    ins_fields = line.split()

                    line = next(f)  # transition line
                    trans_fields = line.split()
                    assert len(trans_fields) == 7

                    node = HmmerNode(
                        match_emit=[cls.str_to_prob(s) for s in match_fields[1:len(model.alph)+1]],
                        ins_emit=[cls.str_to_prob(s) for s in ins_fields],
                        m_to_m=cls.str_to_prob(trans_fields[0]),
                        m_to_i=cls.str_to_prob(trans_fields[1]),
                        m_to_d=cls.str_to_prob(trans_fields[2]),
                        i_to_m=cls.str_to_prob(trans_fields[3]),
                        i_to_i=cls.str_to_prob(trans_fields[4]),
                        d_to_m=cls.str_to_prob(trans_fields[5]),
                        d_to_d=cls.str_to_prob(trans_fields[6]),
                    )
                    model.nodes.append(node)
                break

        model._load_null_model()
        return model

    def _load_null_model(self) -> None:
        self.null_emit = []
        for sym in self.alph:
            self.null_emit.append(SWISSPROT_BG.get(sym, 1.0 / len(self.alph)))

    # State index methods (same formulas as C++)
    def b_idx(self) -> int:
        return 0

    def ix_idx(self, n: int) -> int:
        return 5 * n + 1

    def i_idx(self, n: int) -> int:
        return 5 * n + 2

    def mx_idx(self, n: int) -> int:
        return 5 * n - 2

    def m_idx(self, n: int) -> int:
        return 5 * n - 1

    def d_idx(self, n: int) -> int:
        return 5 * n

    def core_end_idx(self) -> int:
        return 5 * len(self.nodes) + 3

    def n_core_states(self) -> int:
        return 5 * len(self.nodes) + 4

    def calc_match_occupancy(self) -> list[float]:
        M = len(self.nodes)
        mocc = [0.0] * M
        mocc[1] = self.nodes[0].m_to_i + self.nodes[0].m_to_m
        for k in range(2, M):
            mocc[k] = (mocc[k-1] * (self.nodes[k].m_to_m + self.nodes[k].m_to_i)
                        + (1.0 - mocc[k-1]) * self.nodes[k].d_to_m)
        return mocc

    def machine(self, local: bool = True) -> Machine:
        """Build core-only machine (same as C++ HmmerModel::machine)."""
        M = len(self.nodes)
        assert M > 0

        states = [MachineState() for _ in range(self.n_core_states())]

        # B state
        states[self.b_idx()].name = "B"
        if local:
            occ = self.calc_match_occupancy()
            Z = sum(occ[k] * (M - k + 1) for k in range(1, M))
            for k in range(1, M):
                states[self.b_idx()].trans.append(
                    MachineTransition(dest=self.m_idx(k), weight=occ[k] / Z))
        else:
            states[self.b_idx()].trans.append(
                MachineTransition(dest=self.m_idx(1), weight=self.b_to_m1))
            states[self.b_idx()].trans.append(
                MachineTransition(dest=self.i_idx(0), weight=self.b_to_i0))
            states[self.b_idx()].trans.append(
                MachineTransition(dest=self.d_idx(1), weight=self.b_to_d1))

        # I0, Ix0
        states[self.ix_idx(0)].trans.append(
            MachineTransition(dest=self.m_idx(1), weight=self.i0_to_m1))
        states[self.ix_idx(0)].trans.append(
            MachineTransition(dest=self.i_idx(0), weight=self.i0_to_i0))

        for sym_i, sym in enumerate(self.alph):
            states[self.i_idx(0)].trans.append(
                MachineTransition(dest=self.ix_idx(0), output=sym,
                                  weight=self.ins0_emit[sym_i]))

        for n in range(M + 1):
            ns = str(n)
            states[self.i_idx(n)].name = f"I{ns}"
            states[self.ix_idx(n)].name = f"Ix{ns}"

            if n > 0:
                nd = self.nodes[n - 1]
                states[self.m_idx(n)].name = f"M{ns}"
                states[self.mx_idx(n)].name = f"Mx{ns}"
                states[self.d_idx(n)].name = f"D{ns}"

                end = (n == M)

                if end:
                    if not local:
                        states[self.mx_idx(n)].trans.append(
                            MachineTransition(dest=self.core_end_idx(), weight=nd.m_to_m))
                else:
                    states[self.mx_idx(n)].trans.append(
                        MachineTransition(dest=self.m_idx(n + 1), weight=nd.m_to_m))
                states[self.mx_idx(n)].trans.append(
                    MachineTransition(dest=self.i_idx(n), weight=nd.m_to_i))
                if not end:
                    states[self.mx_idx(n)].trans.append(
                        MachineTransition(dest=self.d_idx(n + 1), weight=nd.m_to_d))

                dest_after_i = self.core_end_idx() if end else self.m_idx(n + 1)
                states[self.ix_idx(n)].trans.append(
                    MachineTransition(dest=dest_after_i, weight=nd.i_to_m))
                states[self.ix_idx(n)].trans.append(
                    MachineTransition(dest=self.i_idx(n), weight=nd.i_to_i))

                if end:
                    if not local:
                        states[self.d_idx(n)].trans.append(
                            MachineTransition(dest=self.core_end_idx(), weight=nd.d_to_m))
                else:
                    states[self.d_idx(n)].trans.append(
                        MachineTransition(dest=self.m_idx(n + 1), weight=nd.d_to_m))
                    states[self.d_idx(n)].trans.append(
                        MachineTransition(dest=self.d_idx(n + 1), weight=nd.d_to_d))

                for sym_i, sym in enumerate(self.alph):
                    states[self.m_idx(n)].trans.append(
                        MachineTransition(dest=self.mx_idx(n), output=sym,
                                          weight=nd.match_emit[sym_i]))
                    states[self.i_idx(n)].trans.append(
                        MachineTransition(dest=self.ix_idx(n), output=sym,
                                          weight=nd.ins_emit[sym_i]))

                if local:
                    states[self.m_idx(n)].trans.append(
                        MachineTransition(dest=self.core_end_idx(), weight=1))
                    states[self.d_idx(n)].trans.append(
                        MachineTransition(dest=self.core_end_idx(), weight=1))

        states[self.core_end_idx()].name = "E"
        return Machine(state=states)

    def plan7_machine(self, multihit: bool = False, L: float = 400) -> Machine:
        """Build full Plan7 machine with N/C/J/T flanking states."""
        assert len(self.nodes) > 0
        assert len(self.null_emit) == len(self.alph)

        n_core = self.n_core_states()
        # Plan7 state indices (appended after core)
        n_idx = n_core
        nx_idx = n_core + 1
        p7_b_idx = n_core + 2
        cx_idx = n_core + 3
        c_idx = n_core + 4
        jx_idx = n_core + 5
        j_idx = n_core + 6
        t_idx = n_core + 7
        n_total = n_core + 8

        # Build core in local mode
        core = self.machine(local=True)

        states = [MachineState() for _ in range(n_total)]
        for i in range(n_core):
            states[i] = core.state[i]

        # Move B's transitions to Plan7 Begin state
        states[p7_b_idx] = states[self.b_idx()]
        states[p7_b_idx].name = "B"

        # Repurpose index 0 as S
        states[self.b_idx()] = MachineState(name="S", trans=[
            MachineTransition(dest=nx_idx, weight=1.0)
        ])

        # N-terminal flank
        states[n_idx] = MachineState(name="N", trans=[
            MachineTransition(dest=nx_idx, output=sym, weight=self.null_emit[i])
            for i, sym in enumerate(self.alph)
        ])
        states[nx_idx] = MachineState(name="Nx", trans=[
            MachineTransition(dest=n_idx, weight=L / (L + 1)),
            MachineTransition(dest=p7_b_idx, weight=1.0 / (L + 1)),
        ])

        # E transitions
        if multihit:
            states[self.core_end_idx()].trans.extend([
                MachineTransition(dest=cx_idx, weight=0.5),
                MachineTransition(dest=jx_idx, weight=0.5),
            ])
        else:
            states[self.core_end_idx()].trans.append(
                MachineTransition(dest=cx_idx, weight=1.0))

        # C-terminal flank
        states[c_idx] = MachineState(name="C", trans=[
            MachineTransition(dest=cx_idx, output=sym, weight=self.null_emit[i])
            for i, sym in enumerate(self.alph)
        ])
        states[cx_idx] = MachineState(name="Cx", trans=[
            MachineTransition(dest=c_idx, weight=L / (L + 1)),
            MachineTransition(dest=t_idx, weight=1.0 / (L + 1)),
        ])

        # J loop
        if multihit:
            states[j_idx] = MachineState(name="J", trans=[
                MachineTransition(dest=jx_idx, output=sym, weight=self.null_emit[i])
                for i, sym in enumerate(self.alph)
            ])
            states[jx_idx] = MachineState(name="Jx", trans=[
                MachineTransition(dest=j_idx, weight=L / (L + 1)),
                MachineTransition(dest=p7_b_idx, weight=1.0 / (L + 1)),
            ])
        else:
            states[j_idx] = MachineState(name="J")
            states[jx_idx] = MachineState(name="Jx")

        # Terminal
        states[t_idx] = MachineState(name="T")

        return Machine(state=states)
